Quantizer pass to insert quant-dequant nodes into IR

nishantpdce opened the pull request:
Summary:
- Quantizer pass to mutate IR by inserting quant-dequant nodes
before and after nodes which support quantized ops. This information
will be used by jit compiler to substitute with quantized ops
- This currently covers simple model. It will be expanded later
for subgraph pattern matching to cover more complex patterns

Differential Revision: D14592265

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -0,0 +1,151 @@
+}
+
+// Add Quant-Dequant node pair before quantizable node
+void addQuantDeQuantBefore(torch::jit::Node* def, torch::jit::Value* v) {
```
`addQuantDeQuantAfter` and `addQuantDeQuantBefore` are almost identical, and one of them is almost certainly not needed. Since the entire graph is
N -> V -> N -> V ...
Each V is both input of something and output of something (except values which are external inputs or outputs), and thus quantizing only inputs or only outputs should be sufficient.

nishantpdce replied:
I had tried this approach. Consider this
N->(V1, V2)
V1 -> (N1, N2)
V2 -> (N3, N4)

Now i am interested in inserting quant node between V1->N2 and V2->N4. If i try doing the output only approach. i need to spot at N, scan over all outputs Vi, then scan over all uses of Vi and insert quant node. Also this require having some api like replaiceSpecificUseWith() which is not supported. Instead JIT provides replaceInputWith and replaceOutputWith which operates at node level. So if we are interested in V,N we use replaceInputWith and for N,V we use replaceOutputWith. Please let me know your thoughts

ZolotukhinM replied:
What you described meant that you wanted to quantize inputs of N2 and N4.
What I tried to say is that if you go over all inputs of all nodes in the graph you will visit all values in the graph (except external outputs). Similarly, if you go through all outputs of all nodes, you will also visit all values in the graph (except external inputs). Thus, there is no reason to visit both inputs and outputs - in this case you will visit every value twice (except external inputs and outputs which would need a special handling anyway).

nishantpdce replied:
Let me explain with input only case : 
we will visit only quantizable nodes. if we do just input or output, we will have to visit all nodes and also we might have more quant nodes than needed. Assuming we do inputs only approach, now for this case  Eg V0 -> N -> V1 ->(N1, N2, N3) will look like V0 -> Q -> N -> V1 -> (Q1,N1; Q2,N2; Q3,N3).  We cannot use replaceAll at input because of previous case of N2 and N4 that i described. If this case can be avoided then i agree only input approach is good

If we do both input and output, i agree we will have duplicate node visit in case N1->N2 are both quantizable, but graph will look like : V0 -> Q -> N -> V1 -> Q1->(N1; N2; N3)..

nishantpdce replied:
Actually even for case  where consecutive nodes are quantizable, we will skip the second loop because there is a check for dequant node

ZolotukhinM reviewed (changes requested):
Hi! Please find some remarks inline. It's also hard to review only the relevant pieces, since the PR includes also changes from base commits - please use `ghstack` to fix that, and I'll take another look after that.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -23,7 +72,30 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph) {
+    if (checkIfNodeValuesQuantizable(n) != 0) {
+      for (torch::jit::Value* v : n->outputs()) {
+        if (v->isCompleteTensor()) {
+          addQuantDeQuantNodes(v);
```
We could use `WithInsertPoint`- we won't need to pass the node into `addQuantDeQuantNodes` then. Also we can return the created value from this function and move `replaceInputWith`/`replaceAllUsesWith` out of it - this should make the function simpler.

nishantpdce replied:
WithInsertPoint throws error while mutating the input as for input node v->input belong to different block. replaceInputWith()/replaceAllUsesWith() need to happen before we set v as input to quant. Moreover i feel the logic might scatter outside this api. For now we will retain this as is

nishantpdce replied:
I have slightly modified logic. The special case node inputs will kick in only for external inputs

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -23,7 +72,30 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph) {
+      }
+      // This avoid duplicate insertion of quant-dequant pair for consecutive
+      // quantizable nodes which do not form pattern
+      if (n->prev() != nullptr && n->prev()->kind().toQualString() !=
```
There is no guarantees `n->prev` would be a dequant node even if the value was quantized - we should check for `v->def` instead.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,15 +1,64 @@
+    n = v->node();
+    beforeValue = true;
+  }
+  torch::jit::Node* quant =
```
Nit: `torch::jit::` is not needed.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -23,7 +72,30 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph) {
 
 void InsertFakeQuantNodes(std::shared_ptr<Graph>& graph) {
-  throw std::runtime_error("Pass not implemented yet!");
+  for (const auto& n : graph->nodes()) {
```
Please also add a code for visiting subblocks. It's very straightforward to add and there is no reason for restricting to a single block.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,15 +1,64 @@
+    quant->moveBefore(dequant);
+    n->replaceInputWith(v, dequant->outputs()[0]);
+  }
+  quant->outputs()[0]->setUniqueName(v->uniqueName() + ".quant");
```
Where is  the output added to the quant node?

nishantpdce replied:
as part of create(). Default number of outputs = 1

ZolotukhinM reviewed (changes requested):
Hi! Please find some comments inline. Once they are addressed, I think this PR would be ready to go.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,15 +1,64 @@
+#include <cstdbool>
```
Why do we need this include btw?

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+int checkIfNodeValuesQuantizable(Node* n) {
+  // This is map for quantizable nodes. It will be expanded in future to
+  // support more ops and patterns.
+  static const std::unordered_set<std::string> nodeLookup =
```
Please use the OperatorSet object and not operator names:

```
OperatorSet cannot_propagate_shape_by_running_it = {
      "aten::solve(Tensor self, Tensor A) -> (Tensor, Tensor)",
      "aten::inverse(Tensor self) -> Tensor",
  };
```

This enforces that when someone changes the arguments to an operator, they also must look at the places where it is matched.

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+  return nodeLookup.count(n->kind().toQualString());
+}
+
+bool isTypeTensor(Value *v) {
```
This function is unneeded:

```
v->type()->isSubtypeOf(TensorType::get())
```

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+
+// Check if Value external input to block
+bool checkifExternalInput(Value* v) {
+  return v->node()->kind().toQualString() == std::string("prim::Param");
```
```v->node()->kind() == prim::Param```

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+
+  // Attach inputs to quant and dequant nodes
+  quant->addInput(v);
+  dequant->addInput(quant->outputs()[0]);
```
```quant->output()```

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+  dequant->outputs()[0]->setUniqueName(v->uniqueName() + ".dequant");
+
+  // Add quant-dequant nodes to the graph at appropriate insertion points
+  if (!external) {
```
This can use a comment describing what the before/after graph should look like for both the parameter and non-parameter case. It is hard for me to know if this is correct because I do not know the intent.

I would also suggested splitting this function into a Param and non-Param function, because the way it edits the graph is different for each case.

ZolotukhinM replied:
I agree that splitting the function would help.

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -88,7 +158,37 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+      }
+      // This is to quantize external inputs. Nodes for external input
+      // do not belong to any block
+      for (auto& v : n->inputs()) {
```
Why not loop over the graph/block inputs explicitly instead of finding them as inputs?
Isn't this going to run multiple times on Graph/ block inputs?

nishantpdce replied:
Yeah i considered that approach. For that i need to iterate over v->uses and modify if i identify use as a quantizable node. Documentation said any change to v will invalidate the use iterator. Else  i need to accumulate quantizable uses for v first, and then modify graph. Please suggest

zdevito reviewed (commented):
See attached comments, some comments would help me understand what I am checking for.

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -0,0 +1,151 @@
+  }
+}
+
+void InsertFakeQuantNodes(std::shared_ptr<Graph>& graph) {
```
I will have a following change for that

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1061,7 +1061,34 @@ def observe(x, name):
+                x = F.max_pool2d(x, 2, 2)
+                x = x.view(-1, 4 * 4 * 50)
+                x = F.relu(self.fc1(x))
+                x = self.fc2(x)
```
We don't need a real model to check that quant-dequant nodes are inserted properly, please minimize the test.

ZolotukhinM replied:
Please address this remark.

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1061,7 +1061,34 @@ def observe(x, name):
+                return x
+
+        trace = testModule()
+        torch._C._jit_pass_constant_propagation(trace.graph)
```
Please add a comment explaining why we execute const-prop here.

raghuramank100 replied:
Can we elaborate more? What is the constant prop pass doing? If I only need to know the op (eg relu) whose output I need to quantize, why do I need this pass? Can you clarify in the comments?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+#include <stack>
```
Why has this line moved up?

ZolotukhinM replied:
Please address this remark too.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,84 @@
+namespace {
+// QuantizerUtils
+
+int checkIfNodeValuesQuantizable(Node* n) {
```
This can be `bool` instead of `int`. Also, it could use a better name: when one reads `if(checkSomething)` it's less obvious what the check returns compared to `if(isSomething)` for instance.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -88,7 +158,37 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+      // This is to quantize external inputs. Nodes for external input
+      // do not belong to any block
+      for (auto& v : n->inputs()) {
+        if (checkifExternalInput(v) && isTypeTensor(v)) {
```
Do we only quantize external inputs? What about intermediate values?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -88,7 +158,37 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+      // Basic notion is for a quantizable node to inspect inputs and outputs
+      // We quantize all the outputs
+      // We selectively quantize the inputs depending on parent node
+      for (auto& v : n->outputs()) {
```
Since you use explicit types in other loops in the function (like `Node *` and `Block *`), you could use `Value *` here as well.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -88,7 +158,37 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+        blocks_to_visit.push(subblock);
+      }
+
+      if (!checkIfNodeValuesQuantizable(n)) {
```
This name is a bit misleading - it looks like the function is going to check values, while in fact it only checks the node type and doesn't look at values.

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -991,6 +992,45 @@ def fn(x, y):
+                   .check_next("return")                         \
+                   .run(str(trace))
+
+    def test_expand_insert_fakequant(self):
```
Also, how do we pass qparams here? Should we address that in this patch or do you plan to follow-up on this in a separate PR?

Btw, please fix the testname - it was a copypaste error I originally made (`s/test_expand_insert_fakequant/test_insert_fakequant/`).

nishantpdce replied:
yeah different patch for that

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -88,7 +158,37 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
 }
 
 void InsertFakeQuantNodes(std::shared_ptr<Graph>& graph) {
```
Since we're inserting quant-dequant nodes explicitly (instead of first inserting fake-quants and then expanding them into quant-dequant pairs), please remove `ExpandFakeQuantNodes` function and update the comments in `quantization.h`.

ZolotukhinM reviewed (commented):
I added some comments inline.

nishantpdce commented:
Fixed clang error

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,91 @@
+  Node* dequant =
+      n->owningGraph()->create(at::Symbol::fromQualString("aten::dequant"));
+  AT_ASSERT(dequant != nullptr);
+  dequant->outputs()[0]->setUniqueName(v->uniqueName() + ".dequant");
```
`s/outputs()[0]/output()/`

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -87,8 +164,89 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+  blocks_to_visit.push(graph->block());
+  // For storing quantizable values - node pairs that are external
+  // or intermediate inputs to quantizable nodes
+  std::vector<std::pair<Value*, Node*>> quantInputs;
```
We could use `Use*` instead of `pair<Value*, Node*>`, I believe it would express the logic better.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -87,8 +164,89 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+        //         N1 is not quantizable node. N4 and N7 are quantizable nodes.
+        //         So we cache the (V1, N4) and (V2, N7) as insertion branch
+        //         for quant-dequant nodes
+        for (auto& v : n->outputs()) {
```
Please use `Value *` instead of `auto&`, it's easier to read (here and in other similar places).

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -87,8 +164,89 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+          }
+        }
+      } else {
+        // This iteration is similar to external inputs to block.
```
Why do we do this? Why not just iterate over `inputs`?

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -34,16 +34,17 @@ TORCH_API void InsertObserverNodes(
     Node* observer_node);
 
-/** \brief Inserts fake-quant nodes.
+/** \brief Inserts quant-dequant nodes.
```
Please remove `ExpandFakeQuantNodes` from this file, since we are no longer going the FakeQuant nodes route.

ZolotukhinM reviewed (commented):
Some more comments inline.

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2,13 +2,94 @@
+} // namespace
+
+// PyBind APIs
 void ExpandFakeQuantNodes(std::shared_ptr<Graph>& graph) {
```
Can we remove this pass for now?

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -87,8 +168,80 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+  // Insert the quant-dequant pair for values inputs to quantizable nodes
+  for (auto& ele : quantInputs) {
+    addQuantDeQuantNodesForInput(ele.first, ele.second);
+  }
```
Request adding more tests: There is a lot of logic here in terms of checking for quantizable/non-quantizable nodes and in iterating across sub-blocks.

raghuramank100 reviewed (changes requested):
Please review my comments to make the requested changes.

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -87,8 +168,80 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+  } // end Block traversal
+
+  // Insert the quant-dequant pair for values output from quantizable nodes
+  for (auto& ele : quantOutputs) {
```
Iterating over an unordered set is non-deterministic and will result in different IR every time.

nishantpdce committed:
Quantizer pass to insert quant-dequant nodes into IR (#18446)

Summary:
- Quantizer pass to mutate IR by inserting quant-dequant nodes
before and after nodes which support quantized ops. This information
will be used by jit compiler to substitute with quantized ops

- This currently covers simple model. It will be expanded later
for subgraph pattern matching to cover more complex patterns
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18446

Differential Revision: D14592265

Pulled By: nishantpdce

fbshipit-source-id: 4393e72600f6d8c2c9006c0842d1050a03b9dbc3
